docs/src/test-components-js.md

```diff
@@ -40,6 +40,10 @@ test('event should work', async ({ mount }) => {
 
 Adding Playwright Test to an existing project is easy. Below are the steps to enable Playwright Test for a React, Vue, Svelte or Solid project.
 
+:::note
+If you're developing with [Storybook](https://storybook.js.org), want to test your components in Angular, build them with Webpack/Rspack, utilize NextJS/SvelteKit framework features (links, image/fonts, etc.), or are hitting [known CT limitations](#known-issues-and-limitations), please see [Storybook's Playwright CT integration](https://github.com/storybookjs/playwright-ct).
```

Comment:
Overall, I'm fine with the reference, but we might want to narrow it down a bit and move it all the way to the end. It is really a friendly pointer to a fellow project, not something we want to actively advocate for.

> hitting [known CT limitations](#known-issues-and-limitations)

I don't think any of the limitations are mitigated with your approach, which one did you have in mind? Some of the limitations in our list are out of date, but fundamentally, your integration approach is a strict subset of the stock one, so I'm not sure if it can affect any of the fundamentals.

> utilize NextJS/SvelteKit framework features (links, image/fonts, etc.

I think SvelteKit can work off the stock svelte integration, we should fix it if it does not. And we have NextJS customers testing their front-end components.

Reply:
Hi Pavel, I'm fine to tone down and de-emphasize the reference, and appreciate your goodwill here.

However, while we're chatting here, I'd like to explain why I think CT + Storybook could be a preferable approach for Playwright as your user base grows in breadth and depth

**Limitations.** These cases AFAIK aren’t handled by the current CT transformation:

1. Passing functions that must be executed in the browser or send/receive non-serializable data like JSX elements
2. Importing non-components from component files (which you've already noted in the docs)

**Framework features.** NextJS and SvelteKit have a variety of framework features that are unsupported by CT but are supported by Storybook.

- NextJS examples:
    - [Routing](https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating)
    - [Image](https://nextjs.org/docs/pages/building-your-application/optimizing/images) loading
    - [Font](https://nextjs.org/docs/pages/building-your-application/optimizing/fonts) loading
- SvelteKit examples:
    - Imports that use the special `$lib` alias
    - Components can read environment from `$app/environment`
    - `$app/paths` so you can safely get paths for assets
    - Stores in `$app/stores` are supported out of the box

These integrations are expensive to build and maintain, and it's something that, with our focus on components, we've already signed up to do in Storybook. This might also fit into your plan for Playwright, but if not, it could be a great way to delegate some of those integration complexities.

Reply:
> Passing functions that must be executed in the browser or send/receive non-serializable data like JSX elements

So this bit I still don't understand - I don't think there is any difference between our approach and yours. I.e. out of the box I can have a `.story.tsx` file wrapping the componens under tests and it'll be the exact same thing, or do you do something different to actually allow sending non-serializable data?



> Importing non-components from component files (which you've already noted in the docs)

That's something that we had long time ago, it has long gone now, let me remove it from the list.

> [Routing](https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating)
> [Image](https://nextjs.org/docs/pages/building-your-application/optimizing/images) loading
> [Font](https://nextjs.org/docs/pages/building-your-application/optimizing/fonts) loading

It sounds like you are referring to the Nextjs-specific optimizations rather than the generic support for those - otherwise routing, images and fonts just work in the stock CT story. If that is the case, I agree completely, we don't have a Next.js package, we only have React and I don't think we'll handle those out of the box.
^^ @sand4rt 

> SvelteKit

This one we've been waiting for SvelteKit to stabilize - it uses Vite under the hood, so it would just work, we might need to add a specific package for it to work thought. 
^^ @sand4rt 

For both SvelteKit and Next.js we would like to have stock support - it only takes a couple of files to adapt a framework like SvelteKit, I have no idea what it takes to support Next.js though.

Reply:
> So this bit I still don't understand - I don't think there is any difference between our approach and yours. I.e. out of the box I can have a .story.tsx file wrapping the componens under tests and it'll be the exact same thing, or do you do something different to actually allow sending non-serializable data?

The key difference is that when you specify those things in the `.stories.tsx` file, it's all running in the browser so you don't need to serialize anything:

```js
// MyFunkyWrapper.stories.tsx
export default {
  component: MyFunkyWrapper
}
export const Empty => {
  args: {
     myCallback: undefined,
  }
};
export const WithIcon => {
  args: {
     myCallback: () => <Icon />
  }
};
```

You can then import it into a `.spec.ts` file and as long as you don't try to override the prop in the test, it works:

```js
import * as WrapperStories from './MyFunkyWrapper.stories.tsx';

// works
test('blah', ({mount}) => {
  mount(WrapperStories.WithIcon); 
});
```

However, if you override the prop with a function that does anything weird, it will fail in `@storybook/playwright-ct` (since I copied your trick for marshalling functions back to node):

```
import * as WrapperStories from './MyFunkyWrapper.stories.tsx';

// fails
mount(WrapperStories.Empty, {
  myCallback: () => <Icon />
}); 
```

The point is that in the Storybook version you have a choice about where you want your props to run.

Reply:
I'm sorry, I'm still lost. There is no variation of this code that would work with Storybook here either that would not exist in stock Playwright CT, is that correct?

Is your point that you can stumble upon this foot gun with the stock CT? Because it does not sound like you are removing a 'limitation' in this case, it means that you offer a less capable model by convention where this construct is absent altogether. Which eliminates this failing use case among legit working use cases. Do I get it right?

The reason I bring it up is that we started with the exact same model that you are offering to your users, but based on the overwhelming feedback, we ended up making it more flex (while making it more dangerous). So to us, we are past that learning experience already with the new conscious status quo.

Reply:
The difference is that the story code all runs in the browser, so you can literally pass whatever you want into your component. Thinking about it more, I guess you could also achieve this in Playwright CT by creating a custom wrapper component that you'd use in your tests, rather than using the component you want to test directly, but it feels like an awkward workaround.

I was unaware of the of the history of CT intentionally moving the component setup code out of the browser and into Node. Can you share more about the rationale that led to this decision?

I started this not just as convenient way to reuse Storybook's setup and rendering code, but because I see it as a preferable programming model for browser-based components. You clearly disagree, so I'd like to learn more about how you think about the problem.

Reply:
>  Thinking about it more, I guess you could also achieve this in Playwright CT by creating a custom wrapper component that you'd use in your tests, rather than using the component you want to test directly, but it feels like an awkward workaround.

Yes, exactly! That's what I was trying to say here: https://github.com/microsoft/playwright/pull/26592#discussion_r1305897687. If you name this wrapper `foo.story.tsx` by convention it is no longer that awkward, pretty much indistinguishable from your proposal.

>  Can you share more about the rationale that led to this decision?

It was more or less reactive. I'd say typing was one of the strongest signals. JSX gave us good IDE support when instantiating the components from test. Jest similarity / inertia was another factor.

> I started this not just as convenient way to reuse Storybook's setup and rendering code, but because I see it as a preferable programming model for browser-based components. You clearly disagree, so I'd like to learn more about how you think about the problem.

I never said I don't like a more constrained model, I think it makes a lot of sense. Playwright CT is literally experimental and new, so we could afford using its status to experiment and see what works and what does not. (Maybe unlike Storybook), we aren't opinionated in terms of the component organization and one can literally use the model that you are suggesting with your integration, in Playwright CT, on the level of convention (`foo.story.tsx`).

We are still looking for the right balance in handling the communication between the browser and node. We started with what you had and it was met with a slight pushback from the customers. We are flexible, we decided to try the more risky route (into what we have today). At least non-storybook customers seem to be more happy with the new status quo, but by no means it is final, we still think we can do better to clearly define the runtime(s) boundaries.